docs/rfcs/016-connection-routing.md

```diff
@@ -0,0 +1,150 @@
+# Dispatching a connection
+
+For each client connection, zenith needs to authenticate the
+connection, and route it to the right PostgreSQL instance.
+
+## Authentication
+
+There are three different ways to authenticate:
+
+- anonymous; no authentication needed
+- PostgreSQL authentication
+- github single sign-on using browser
+
+In anonymous access, the user doesn't need to perform any
+authentication at all. This can be used e.g. in interactive PostgreSQL
+documentation, allowing you to run the examples very quickly. Similar
+to sqlfiddle.com.
+
+PostgreSQL authentication works the same as always. All the different
+PostgreSQL authentication options like SCRAM, kerberos, etc. are
+available. [1]
+
+The third option is to authenticate with github single sign-on. When
+you open the connection in psql, you get a link that you open with
+your browser. Opening the link redirects you to github authentication,
+and lets the connection to proceed. This is also known as "Quick
+connect" [2].
+
+
+## Routing the connection
+
+When a client starts a connection, it needs to be routed to the
+correct PostgreSQL instance. Routing can be done by the proxy, acting
+as a man-in-the-middle, or the connection can be routed at the network
+level based on the hostname or IP address.
+
+Either way, Zenith needs to identify which PostgreSQL instance the
+connection should be routed to. If the instance is not already
+running, it needs to be started. Some connections always require a new
+PostgreSQL instance to be created, e.g. if you want to run a one-off
+query against a particular point-in-time.
+
+The PostgreSQL instance is identified by:
+- zenith account (possibly anonymous)
+- cluster (known as tenant in the storage?)
+- branch or snapshot name
+- timestamp (PITR)
+- primary or read-replica
+- one-off read replica
+- one-off writeable branch
+
+When you are using regular PostgreSQL authentication or anonymous
+access, the connection URL needs to contain all the information needed
+for the routing. With github single sign-on, the browser is involved
+and some details - the zenith account in particular - can be deduced
+from the authentication exchange.
+
+There are three methods for identifying the PostgreSQL instance:
+
+- Browser interaction (quick connect)
+- Options in the connection URL and the domain name
+- A pre-defined endpoint, identified by domain name or IP address
+
+### Quick connect
+
+    postgres://<username>@start.zenith.tech/<dbname>
+
+This gives you a link that you open in browser. Clicking the link
```

Comment:
I can't open postgres:// -links in my browser. 

Did you mean that _connecting to_ `postgres://<username>@start.zenith.tech/<dbname>` with psql would give you a link to the browser?

Reply:
Can we pass a github auth token instead? psql -> browser switch sounds bit odd.

Reply:
> Can we pass a github auth token instead? psql -> browser switch sounds bit odd.

Have you tried how it works now? Try connecting with `psql -h start.zenith.tech`. It's a cool party trick to onboard new users, to get them to try it out.

In the UI you can create an authentication token for each cluster and use that instead.

Hmm, one alternative might be to create a one-time token in the web UI and use that to connect. So you would click on a cluster in the we UI, and it would spit out a command line like `psql -h <authentication-token>.start.zenith.tech` that is valid only for a few minutes. That would work with tools other than psql too. The current browser-link system requires that the client displays the NOTICE message that contains the link, but many tools would just ignore them.

Reply:
> `psql -h start.zenith.tech`

I created a token in console and set it in pgpass.

> 
> Hmm, one alternative might be to create a one-time token in the web UI and use that to connect. So you would click on a cluster in the we UI, and it would spit out a command line like `psql -h <authentication-token>.start.zenith.tech` that is valid only for a few minutes. That would work with tools other than psql too. The current browser-link system requires that the client displays the NOTICE message that contains the link, but many tools would just ignore them.

Not sure, token leaks in dns lookups and there's a window to hijack the cluster. That applies to other CNAMEs discussed here too. ie.  customers-unicorn.db.zenith.tech would be very revealing.
What's the expected work flow scripts or periodic cron jobs? wouldn't it also cover one time use as well?

Reply:
> > `psql -h start.zenith.tech`
> 
> I created a token in console and set it in pgpass.
> 
> > Hmm, one alternative might be to create a one-time token in the web UI and use that to connect. So you would click on a cluster in the we UI, and it would spit out a command line like `psql -h <authentication-token>.start.zenith.tech` that is valid only for a few minutes. That would work with tools other than psql too. The current browser-link system requires that the client displays the NOTICE message that contains the link, but many tools would just ignore them.
> 
> Not sure, token leaks in dns lookups and there's a window to hijack the cluster.

Good point.

> That applies to other CNAMEs discussed here too. ie. customers-unicorn.db.zenith.tech would be very revealing.

Hmm. I think we would normally use randomly assigned cluster ids in subdomains, so it would be something like `cluster-1234.db.zenith.tech`.

> What's the expected work flow scripts or periodic cron jobs? wouldn't it also cover one time use as well?

Not sure. You could use the connection URL format, and provide any options you need in the connection URL. Or you could create an endpoint using the API in your script, and connect to that.

Reply:
> Hmm. I think we would normally use randomly assigned cluster ids in subdomains, so it would be something like `cluster-1234.db.zenith.tech`.

Sure, API allows cluster creation with user assigned names. Injecting those strings into CNAMES is a bad idea.